src/compute/src/server.rs

```diff
@@ -417,23 +416,45 @@ pub async fn compute_node_serve(
             })
             .add_service(MonitorServiceServer::new(monitor_srv))
             .add_service(ConfigServiceServer::new(config_srv))
-            .add_service(HealthServer::new(health_srv))
-            .serve_with_shutdown(listen_addr, async move {
-                tokio::select! {
-                    _ = tokio::signal::ctrl_c() => {},
-                    _ = &mut shutdown_recv => {
-                        for (join_handle, shutdown_sender) in sub_tasks {
-                            if let Err(err) = shutdown_sender.send(()) {
-                                tracing::warn!("Failed to send shutdown: {:?}", err);
-                                continue;
-                            }
-                            if let Err(err) = join_handle.await {
-                                tracing::warn!("Failed to join shutdown: {:?}", err);
-                            }
+            .add_service(HealthServer::new(health_srv));
+        let shutdown_signal = async move {
+            tokio::select! {
+                _ = tokio::signal::ctrl_c() => {},
+                _ = &mut shutdown_recv => {
+                    for (join_handle, shutdown_sender) in sub_tasks {
+                        if let Err(err) = shutdown_sender.send(()) {
+                            tracing::warn!("Failed to send shutdown: {:?}", err);
+                            continue;
                         }
-                    },
-                }
-            })
+                        if let Err(err) = join_handle.await {
+                            tracing::warn!("Failed to join shutdown: {:?}", err);
+                        }
+                    }
+                },
+            }
+        };
+        #[cfg(not(madsim))]
+        {
+            use risingwave_common::monitor::connection::{monitored_tcp_incoming, TcpConfig};
+            server
+                .serve_with_incoming_shutdown(
+                    monitored_tcp_incoming(
+                        listen_addr,
+                        "grpc-compute-node-service",
+                        TcpConfig {
+                            tcp_nodelay: true,
+                            keepalive_duration: None,
+                        },
+                    )
+                    .unwrap(),
+                    shutdown_signal,
+                )
+                .await
+                .unwrap();
+        }
+        #[cfg(madsim)]
+        server
+            .serve_with_shutdown(listen_addr, shutdown_signal)
             .await
             .unwrap();
```

Comment:
What about extracting this into an extension trait on `Router`, with `easy_ext::ext`?

Reply:
Great idea! I have changed it.